File: code/python/213.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums)==0: return 0
        if len(nums)<=2: return max(nums)
        def robMax(arr,start,end):
            dp_i_0 = 0
            dp_i_1 = 0
            temp = 0
            for i in range(start,end):
                temp = dp_i_1
                dp_i_1 = max(dp_i_0,dp_i_1)
                dp_i_0 = temp+nums[i]
            return max(dp_i_0,dp_i_1)
        n = len(nums)
        return max(robMax(nums,0,n-1),robMax(nums,1,n))

File: code/java/213.java
class Solution {
    public int robMax(int[] nums,int start,int end){
        int dp_i_0 = 0; // 偷
        int dp_i_1 = 0; // 不偷
        int dp_i = 0;
        for(int i=start;i<end;i++){
            dp_i = dp_i_1;
            dp_i_1 = Math.max(dp_i_0,dp_i_1);
            dp_i_0 = dp_i+nums[i];
        }
        return Math.max(dp_i_0,dp_i_1);
    }

    public int rob(int[] nums) {
        if(nums.length==0) return 0;
        if(nums.length==1) return nums[0];
        if(nums.length==2) return Math.max(nums[0],nums[1]);
        int n = nums.length;
        return Math.max(robMax(nums,0,n-1),robMax(nums,1,n));
    }
}